# Artificial Neural Network (Tensorflow)
# To Predict number of customers about to leave/ close account in a bank
# Dataset - Churn Modelling of a Bank (Kaggle)
## ANN -Fully connected Neural Network with fully connected layers
## ANN can be used for Both Regression and Classification and here we will use for binary classification

# Downlaoding TensorFlow (Inbulit with Keras)

In [1]:
!pip install tensorflow

# Importing the Libraries

In [2]:
import numpy as np
import pandas as pd
import tensorflow as tf #Keras integrated with Tensorflow
print(tf.__version__)

2.16.1


# Part -1 : Data Preprocessing

## Importing the Dataset 

In [3]:
dataset=pd.read_csv(r'E:\7. Deep Learning\Artificial Neural Networks (ANN)\Churn_Modelling.csv')
x=dataset.iloc[:, 3:-1].values
y=dataset.iloc[:, -1].values #[:-All Rows, 3/:-All the columns]
print(x)
print(y)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]
[1 0 1 ... 1 1 0]


## Encoding Categorical Data 
### Label Encoding "Gender"(0/1) column

In [4]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
x[:, 2]=le.fit_transform(x[:, 2])
print(x)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


### One Hot Encoding the " Geography" Column

In [5]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct=ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
x=np.array(ct.fit_transform(x))
print(x) 
#France - 101
#Spain - 100
#Germany - 010

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


### Splitting Dataset in Test Set and Training Set

In [6]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)

## Feature Scaling - VERY VERY Important for Deep Learning
### Whenever we create a neural network feature scaling is absolutely necessary
### APPLY THEM TO ALL OUR FEATURES 

In [7]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
x_train=sc.fit_transform(x_train)
x_test=sc.transform(x_test)

## Part -2 Building the ANN

### Intialize ANN as sequence of layers
#### Create a variable of ann for sequence of layer

In [8]:
ann=tf.keras.models.Sequential() #since keras is integrated in tensorflow this is how we call it\

### Adding Input Layer and First Hidden Layer
#### We use Dense class to add an layer 

In [9]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu')) #units - No. of Neurons in hidden layer have to experiment with it
#this is a shallow neural network 

### Add Second Hidden Layer 

In [10]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu')) #relu - rectifier actiavtion function

### Adding an Output Layer - will change the hyperparameters

In [11]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid')) #sigmoid - activation function for output layer to get probablity
#activation = softmax - for classification of multiple Categories

## Part -3 Training the ANN

### Compile the ANN with paramters :  Optimizer,Loss Function & a Metric
#### loss - way to compute diffrence btw prediction and real results

In [12]:
ann.compile(optimizer = 'adam', loss= 'binary_crossentropy', metrics= ['accuracy']) #'adam'- Name for Stochastic Gradient Descent 
#loss= categorical_crossentropy - for classification of multiple categories 

### Traning the ANN on the Training Dataset

In [13]:
ann.fit(x_train,y_train, batch_size=32, epochs=100)
#batch_size = no of predictions you want to have in the batch to be comapared to the same number of real result
#classic value for batch -32
#epochs - a neural network has to be trained over certain no of epochs so as to improve the accuracy over time
#epochs - 100 (choose any bit it shoudlnt be small)

Epoch 1/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7954 - loss: 0.5674
Epoch 2/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7953 - loss: 0.4949
Epoch 3/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8026 - loss: 0.4436
Epoch 4/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7970 - loss: 0.4352
Epoch 5/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8040 - loss: 0.4300
Epoch 6/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8140 - loss: 0.4125
Epoch 7/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8114 - loss: 0.4273
Epoch 8/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8214 - loss: 0.4150
Epoch 9/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8209 - loss: 0.4119
Epoch 10/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8285 - loss: 0.4064
Epoch 11/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8354 - loss: 0.3895
Epoch 12/100
250/250 ━━━━━━━━━━━━━━━━━━━━

## Part -4 Making Predictions and Evaluating the Model

### Predicting a single observation Result 

In [14]:
print(ann.predict(sc.transform([[1, 0 , 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]]))>0.5)
#sc= since we did scaling we apply and it 
#no - fit - because we dont want the meadna dn median and avaoid information leakage 
#we will get a probability - since we use sigmoid 

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 107ms/step
[[False]]


### Predicting Test Set Result

In [15]:
y_pred=ann.predict(x_test)
y_pred=(y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred), 1), y_test.reshape(len(y_test), 1)), 1))

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


## Making Confusion Matrix

In [16]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm=confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1516   79]
 [ 200  205]]


0.8605

### Stays in Bank = 1516 - Correct Predictions ; 79 - Incorrect Prediction 
### Leaves the Bank = 200 Correct Predictions ; 205 - Incorrect predictions
### Accuracy - 86.05%